src/main/java/net/fabricmc/tinyremapper/AsmClassRemapper.java

```diff
@@ -125,14 +126,26 @@ private static int getLvIndex(String desc, boolean isStatic, int asmIndex) {
 		}
 
 		@Override
-		public void visitParameter(String name, int access) {
-			name = ((AsmRemapper) remapper).mapMethodArg(this.owner, this.name, this.desc, getLvIndex(this.desc, isStatic, argsVisited), name);
-			argsVisited++;
-			super.visitParameter(name, access);
+		public void visitParameter(final String name, int access) {
+			String newName = ((AsmRemapper) remapper).mapMethodArg(this.owner, this.name, this.desc, getLvIndex(this.desc, isStatic, argsVisited), name);
+
+			if (newName == null || newName.equals(name)) {
+				newName = renamedInvalidLocals.get(argsVisited);
+			}
+
+			if(this.isAbstract && (newName == null || newName.equals(name))) {
+				newName = "p_" + nameCounts.compute("p", (k, v) -> (v == null) ? 1 : v + 1);
```

Comment:
The type of a parameter can be inferred from the method description as per [the spec](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.24):
> The i'th entry in the parameters array corresponds to the i'th parameter descriptor in the enclosing method's descriptor.

Taking the assumption that `argsVisited` is never erroneously bumped more (or indeed less) than necessary, that provides `i` in `Type.getArgumentTypes(this.desc)` which could then be handled as `#visitLocalVariable` does.

Restricting this to abstract methods only is slightly questionable too as an LVT doesn't have to be present on a normal method (as Mojang indeed didn't until 1.8.4). There won't be a direct way to know whether a method does or doesn't have one until `#visitEnd` is reached as at that point all the method information that the class has will have been passed through. Perhaps it is best that `#checkParameters` is only called once from `#visitEnd` so that any parameters names from the LVT have definitely been passed?

Reply:
> The type of a parameter can be inferred from the method description
Already did that in a newer commit. Thanks for pointing it out anyways :)

> Restricting this to abstract methods only is slightly questionable too as an LVT doesn't have to be present on a normal method (as Mojang indeed didn't until 1.8.4). There won't be a direct way to know whether a method does or doesn't have one until `#visitEnd` is reached as at that point all the method information that the class has will have been passed through. Perhaps it is best that `#checkParameters` is only called once from `#visitEnd` so that any parameters names from the LVT have definitely been passed?
Sure, I could do that.